### PR TITLE
Fix unknown warning "-Wformat-truncation" build failure for arm 

### PR DESCRIPTION
**Pull request by gwang-msft**
**Description**: Fix unknow warning "-Wformat-truncation" build failure for arm 

**Motivation and Context**
- #8701 add ignore warning "-Wformat-truncation" for all compilers, where "-Wformat-truncation" is not defined for clang, this will cause arm build failure or warning if being built by clang, such as Android, M1 mac and iOS
- The fix is to check if the warning is defined before ignoring it

### CODE REVIEW

File: onnxruntime/test/providers/kernel_def_hash_test.cc [hunk 1]
```diff
@@ -63,8 +63,10 @@
 #pragma warning(disable : 28020)
 #elif __aarch64__
 #pragma GCC diagnostic push
+#if !defined(__has_warning) || __has_warning("-Wformat-truncation")
```
Comment (edgchen1) [inline onnxruntime/test/providers/kernel_def_hash_test.cc:66]: if __has_warning is not defined, is the rest of the line valid?

Comment (gwang-msft) [reply onnxruntime/test/providers/kernel_def_hash_test.cc:66]: Yes, __has_warning  seems to be only defined for clang. gcc doesn't have it, as we need to ignore warning if using gcc

Comment (snnn) [reply onnxruntime/test/providers/kernel_def_hash_test.cc:66]: In https://github.com/microsoft/onnxruntime/blob/master/cmake/CMakeLists.txt#L917
Add  
```
 check_cxx_compiler_flag(-Wformat-truncation HAS_FORMAT_TRUNCATION)
```
Then open https://github.com/microsoft/onnxruntime/blob/master/cmake/onnxruntime_config.h.in, add HAS_FORMAT_TRUNCATION there.

Then in this file, 
```c++
#include <onnxruntime_config.h>

#ifdef HAS_FORMAT_TRUNCATION 
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wformat-truncation"
#endif
```

**Review by edgchen1 — APPROVED** [review]

File: onnxruntime/test/providers/kernel_def_hash_test.cc (hunk 1 above)
Comment (gwang-msft) [reply onnxruntime/test/providers/kernel_def_hash_test.cc]: Yep, I think this is simpler

**Review by jywu-msft — APPROVED** [review]

**Comment by jywu-msft** [comment]
thanks for fixing!

**Pull request merged**